cirq/linalg/distance_measures.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Distance measures between quantum states and operations."""
+
+import numpy as np
+import scipy
+
+
+def _sqrt_positive_semidefinite_matrix(mat: np.ndarray) -> np.ndarray:
+    """Square root of a positive semidefinite matrix."""
+    eigs, vecs = scipy.linalg.eigh(mat)
+    # Zero out small negative entries
+    eigs = np.maximum(eigs, np.zeros(eigs.shape, dtype=eigs.dtype))
+    return vecs @ np.diag(np.sqrt(eigs)) @ vecs.T.conj()
+
+
+def fidelity(state1: np.ndarray, state2: np.ndarray) -> float:
```

Comment:
`STATE_VECTOR_LIKE` doesn't encompass density matrices though.

Reply:
Indeed. As discussed in the meeting, please define some way to represent quantum states (both pure and mixed) in cirq and then use it here. Note that the new representation (which can probably just be a union type) should encompass STATE_VECTOR_LIKE. Consequently, this function will need to be slightly generalized to handle other representations of pure states than just numpy arrays of amplitudes (for example, when both states are given as indices into the computational basis then the fidelity is simply 1 if the indices are equal and 0 otherwise).

(If you prefer you can do this in a separate PR.)

Reply:
I'd be happy to merge this initial version and then generalize it to work on the other understood values. It might get a bit tricky in the general case, because density matrices look quite similar to the state tensor of a pair of qudits.

Reply:
Fine with deferral, too.

@kevinsung Could you file an issue so we remember to do this later?